src/persist/src/operators/source.rs

```diff
@@ -149,21 +210,92 @@ mod tests {
                     .expect("write was successful");
             }
             write.seal(6).recv().expect("seal was successful");
-            recv
+            (oks, errs)
         });
 
-        let mut actual = recv
+        assert_eq!(
+            errs.extract()
+                .into_iter()
+                .flat_map(|(_time, data)| data.into_iter().map(|(err, _ts, _diff)| err))
+                .collect::<Vec<_>>(),
+            Vec::<String>::new()
+        );
+
+        let mut actual = oks
             .extract()
             .into_iter()
             .flat_map(|(_, xs)| xs.into_iter().map(|((k, _), _, _)| k))
             .collect::<Vec<_>>();
         actual.sort();
+
         let expected = (1usize..=5usize).map(|x| x.to_string()).collect::<Vec<_>>();
         assert_eq!(actual, expected);
 
         Ok(())
     }
 
+    /// Verifies that `persisted_source()` correctly downgrades to the upper frontier
+    /// of already existing/sealed data when it is re-created based on an existing collection.
+    ///
+    /// There is no easy way to assert on a probe. Instead, we step the worker until the probe is
+    /// no longer `less_than` the expected timestamp. This test would hang if the source were not
+    /// working as expected.
+    #[test]
+    fn eager_capability_downgrade() -> Result<(), Error> {
+        let temp_dir = tempfile::tempdir().expect("failed to create temp directory");
+
+        // We're using a file-based runtime because we cannot use a memory-based runtime to
+        // simulate shutdown and restart.
+        fn create_runtime(base_path: &Path) -> Result<RuntimeClient, Error> {
+            let log_dir = base_path.join("log");
+            let blob_dir = base_path.join("blob");
+            let lock_info = LockInfo::new("lock".to_owned(), "no_details".to_owned())?;
+            let runtime = runtime::start(
+                FileLog::new(log_dir, lock_info.clone())?,
+                FileBlob::new(blob_dir, lock_info)?,
+                &MetricsRegistry::new(),
+            )?;
+            Ok(runtime)
+        }
+
+        let seal_ts = {
+            let p = create_runtime(temp_dir.path())?;
+
+            let (write, _) = p.create_or_load("1").unwrap();
+            for i in 1..=5 {
+                write
+                    .write(&[((i.to_string(), ()), i, 1)])
+                    .recv()
+                    .expect("write was successful");
+            }
+            write.seal(6).recv().expect("seal was successful");
+            6
+        };
+
+        let guards = timely::execute(Config::thread(), move |worker| {
```

Comment:
You're right! I thought I needed `execute()` because I need manuall control over stepping, but it also works with `execute_directly()` because returning from the closure will shut down the sources, meaning the `step()` inside `execute_directly()` will also finish.